Add non-blocking timer module

Nemo157 opened the pull request:
As I mentioned in https://github.com/japaric/embedded-hal/issues/59 the `Duration` struct gets compiled out fully in my test code, but I'm still not certain using it is the best idea in case it doesn't.

therealprof commented on src/timer.rs:
```diff
@@ -0,0 +1,48 @@
+
+    fn start<T>(&mut self, count: T) where T: Into<Self::Time> {
+        let duration = count.into();
+        debug_assert!(duration.as_secs() < ((u32::MAX as u64 - duration.subsec_micros) / 1_000_000));
```
I really don't like `debug_assert!()` calls in embedded because the chances are high that the code will be never exercised.

therealprof replied:
```
error[E0615]: attempted to take value of method `subsec_micros` on type `core::time::Duration`
  --> src/timer.rs:27:73
   |
27 |         debug_assert!(duration.as_secs() < ((u32::MAX as u64 - duration.subsec_micros) / 1_000_000));
   |                                                                         ^^^^^^^^^^^^^
   |
   = help: maybe a `()` to call it is missing?
```

Nemo157 replied:
Yeah, I dislike the `debug_assert` call too. The alternatives I see are:

 * `assert` and hope it gets optimised out
    * I feel like this isn't too bad a choice, since using `Time = Duration` already relies on `CountDown::start` being inlined and optimised it's highly likely this will be ok.
 * don't assert at all, just have it as a documentation item
    * since this is likely to be passed into places that aren't `nrf51` specific they won't see this documentation and it's likely to go wrong at some point
 * Figure out some way to do constrain it at compile time
    * I don't think even once const parameters are stable that Rust will be capable of doing this ergonomically

So, if you're happy with it I'll change this to a normal `assert`

therealprof replied:
A normal `assert` is fine.

Nemo157 commented:
Fixed compile error (I had switched a constant 1 second into the `subsec_micros` call without actually compiling it 🤦‍♂️), switched `debug_assert` -> `assert` and ran rustfmt.

I'll try and push an example tomorrow, unfortunately it'll be going through a `embedded-hal` -> `futures` translation layer so might not be the easiest for you to see what's happening, but should give you an idea.

Nemo157 commented:
I've pushed an example at https://github.com/Nemo157/nrf51-blinker-rs, that uses this + some even more experimental GPIOTE edge detection to wait for a timer to expire or button to be pressed non-blockingly.

therealprof commented:
Okay, that example is rather weird but it works. :-D I'm somewhat curious where you implemented the GPIOTE stuff you're talking about...

therealprof commented:
NB: I've just pushed out a new version 0.4.2 to crates.io. Thanks a lot for your contributions.

Nemo157 commented:
https://github.com/Nemo157/nrf51-hal/compare/56d7cc66b25a46c17b6598131fa3184e1430752c...input-edges

I really don’t like the API I added to embedded-hal for it though; I’m going to try and come up with a better design before I’ll be opening PRs to add it. Also want to come up with a compile time checked API for assigning the event channels to avoid the assert.
